File: tools/generate_setlist_test_matrix.cpp
// Standalone hardware-validation CLI tool -- NOT part of the shipped app,
// not linked into kronos_editor. Loads a real .PCG/.SNG file, writes a
// matrix of Setlist Color/Volume/Comment/Font-size test permutations into
// Set List 0's slots 010-045 (leaving slot 000, the source entry, and
// everything else in the file untouched), and saves the result next to the
// input as "<name>-test<ext>" -- so the output can be loaded onto a real
// Kronos and checked by eye, slot by slot. See STATE.md for the full
// rationale: this is how every confirmed byte offset in this project has
// been verified -- real ground truth, not a plausible-looking guess.
//
// Slots 025-029 (group 4) specifically probe word-wrap: the same sequential
// numbered-token Comment text ("01 02 03 ... 80") at each of the 5 Font
// sizes, so wherever the real hardware breaks each line can be read
// straight off the screen and reported back exactly -- see
// frontend/readme-screen.txt's own (unverified, likely fabricated) claims
// about Kronos text rendering for why this needed a real check rather than
// trusting a plausible-looking guess. Confirmed 2026-08-06, see STATE.md.
//
// Slots 030-045 (group 5) probe Color -- one slot per real Kronos Set List
// color, in pane.js's SETLIST_COLOR_NAMES order (that order is itself not
// independently confirmed against real hardware yet -- this group doubles
// as that check).
//
// tools/generate_setlist_test_matrix.js is the equivalent devtools-console
// version -- same idea, driven through the real running app's bridge
// instead of this standalone binary. Use whichever is more convenient:
// this tool needs no GUI/WebView at all (just `cmake --build` and run it
// from a terminal), the JS version is better suited to interactive,
// one-off bridge testing inside the real running app.
//
// Deliberately built against ONLY PcgFile's public raw-byte API
// (songRecordBytes()/putSongRecordBytes()/save()) -- same minimal
// dependency set as tests/pcg_file_test.cpp (no main.cpp, no
// EditorBridge.cpp, no CHOC/WebView), so this builds and runs in seconds
// with no platform WebView toolchain needed.
//
// The Comment/Font-size/Volume ENCODING below mirrors -- byte for byte --
// src/kronos/PcgFile.cpp's private kSbk* constants (used by readSlotParams())
// and frontend/components/kronos/setlist-editor-comment-and-font.js's/
// setlist-editor-color.js's/setlist-editor-volume.js's JS codecs. It's
// necessarily a third expression of the same confirmed
// encoding: PcgFile.cpp's masks are private to that translation unit (by
// design -- nothing outside it needs Kronos byte offsets), and this tool
// has no WebView to run the JS codecs in. If the confirmed encoding for any
// of these fields ever changes, this needs to change with it.
//
// Usage: generate_setlist_test_matrix <input.pcg>
//   Writes <input-without-extension>-test<ext> next to it.

#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>

#include "kronos/PcgFile.h"

namespace {

constexpr int kSetlistIndex = 0;
constexpr int kSourceSlot = 0;

// Mirrors PcgFile.cpp's kSbk* constants exactly (see file header comment).
constexpr size_t kTypeColorOffset = 12;
constexpr uint8_t kFontSizeLowMask = 0xC0;
constexpr uint8_t kColorMask = 0x3C;  // bits 2-5 of the same byte as Font size's low bits
constexpr size_t kVolumeOffset = 16;
constexpr size_t kFontTransposeOffset = 17;
constexpr uint8_t kFontSizeHighMask = 0x10;
constexpr size_t kCommentOffset = 18;
constexpr size_t kRecordSize = 542;

// Display order (ascending small-to-large) paired with each size's
// confirmed encoding value (docs/content/format/index.md §4.4 -- 0=S is the true
// baseline, not first alphabetically/by size; matches
// kronos::FontSize's own enum order exactly, S=0..XL=4).
const char* const kFontSizeNames[5] = {"XS", "S", "M", "L", "XL"};
constexpr int kFontSizeValues[5] = {1, 0, 2, 3, 4};
constexpr int kVolumes[5] = {0, 1, 10, 100, 127};

// Same order as pane.js's SETLIST_COLOR_NAMES -- confirmed against real
// hardware via this very group, see STATE.md.
const char* const kColorNames[16] = {
    "Default", "Charcoal", "Brick",   "Burgundy", "Ivy",    "Olive", "Gold",  "Cacao",
    "Indigo",  "Navy",     "Rose",    "Lavender", "Azure",  "Denim", "Silver", "Slate",
};

// Group 4's word-wrap probe text: "01 02 03 ... 80", built once in main().
std::string makeWrapTestText() {
    std::string text;
    for (int i = 1; i <= 80; ++i) {
        if (i > 1) text += ' ';
        if (i < 10) text += '0';
        text += std::to_string(i);
    }
    return text;
}

// Mirrors setlist-editor-comment-and-font.js's encodeSetlistComment() --
// masked read-modify-write so this never touches Color/Transpose/the
// still-unexplained bits sharing these same two bytes.
void encodeComment(std::vector<uint8_t>& bytes, const std::string& comment, int fontSizeValue) {
    const auto lowBits = static_cast<uint8_t>(((fontSizeValue & 2) ? 0x80 : 0) | ((fontSizeValue & 1) ? 0x40 : 0));
    bytes[kTypeColorOffset] = static_cast<uint8_t>((bytes[kTypeColorOffset] & ~kFontSizeLowMask) | lowBits);
    bytes[kFontTransposeOffset] = static_cast<uint8_t>((bytes[kFontTransposeOffset] & ~kFontSizeHighMask) |
                                                         ((fontSizeValue & 4) ? kFontSizeHighMask : 0));

    for (size_t i = kCommentOffset; i < kRecordSize; ++i) bytes[i] = 0;
    for (size_t i = 0; i < comment.size() && kCommentOffset + i < kRecordSize - 1; ++i) {
        bytes[kCommentOffset + i] = static_cast<uint8_t>(comment[i]);
    }
}

// Mirrors setlist-editor-volume.js's encodeSlotVolume() -- a plain byte,
// no masking needed (not shared with any other field).
void encodeVolume(std::vector<uint8_t>& bytes, int volume) {
    bytes[kVolumeOffset] = static_cast<uint8_t>(volume < 0 ? 0 : (volume > 127 ? 127 : volume));
}

// Mirrors setlist-editor-color.js's encodeSlotColor() -- masked read-modify-
// write, same discipline as encodeComment()'s Font size handling (this byte
// is shared with isProgram/Font size too). `color` is 1-based (1..16).
void encodeColor(std::vector<uint8_t>& bytes, int color) {
    const int clamped = color < 1 ? 1 : (color > 16 ? 16 : color);
    const auto fieldBits = static_cast<uint8_t>(((clamped - 1) << 2) & kColorMask);
    bytes[kTypeColorOffset] = static_cast<uint8_t>((bytes[kTypeColorOffset] & ~kColorMask) | fieldBits);
}

// `color` < 0 means "leave the source's own Color untouched" -- same
// sentinel convention as fontSizeValue/volume being pre-resolved to the
// source's own value at each call site below when a group doesn't vary
// that field.
bool writeSlot(kronos::PcgFile& pcg, int slot, const std::vector<uint8_t>& sourceBytes, int fontSizeValue,
               int volume, int color, const std::string& label, std::string& error) {
    std::vector<uint8_t> bytes = sourceBytes;
    encodeComment(bytes, label, fontSizeValue);
    encodeVolume(bytes, volume);
    if (color >= 0) encodeColor(bytes, color);
    if (!pcg.putSongRecordBytes(kSetlistIndex, slot, bytes)) {
        error = "putSongRecordBytes() failed for slot " + std::to_string(slot);
        return false;
    }
    std::printf("Slot %d: %s\n", slot, label.c_str());
    return true;
}

// "<dir>/<name>.pcg" -> "<dir>/<name>-test.pcg" -- keeps whatever extension
// the input actually has (".PCG"/".SNG" are both real dialects this format
// covers, see docs/content/format/index.md) rather than assuming ".pcg" specifically.
std::string testOutputPath(const std::string& inputPath) {
    size_t dot = inputPath.find_last_of('.');
    size_t slash = inputPath.find_last_of("/\\");
    if (dot != std::string::npos && (slash == std::string::npos || dot > slash)) {
        return inputPath.substr(0, dot) + "-test" + inputPath.substr(dot);
    }
    return inputPath + "-test";
}

}  // namespace

int main(int argc, char** argv) {
    if (argc != 2) {
        std::fprintf(stderr, "Usage: %s <input.pcg>\n", argv[0]);
        return 1;
    }

    const std::string inputPath = argv[1];
    kronos::PcgFile pcg;
    std::string error;
    if (!pcg.load(inputPath, error)) {
        std::fprintf(stderr, "Failed to load %s: %s\n", inputPath.c_str(), error.c_str());
        return 1;
    }

    if (static_cast<int>(pcg.setlists().size()) <= kSetlistIndex ||
        static_cast<int>(pcg.setlists()[kSetlistIndex].songs.size()) <= kSourceSlot) {
        std::fprintf(stderr, "Set List %d / slot %d doesn't exist in this file\n", kSetlistIndex, kSourceSlot);
        return 1;
    }

    auto sourceBytesOpt = pcg.songRecordBytes(kSetlistIndex, kSourceSlot);
    if (!sourceBytesOpt) {
        std::fprintf(stderr, "No SBK1 record at Set List %d / slot %d\n", kSetlistIndex, kSourceSlot);
        return 1;
    }
    const std::vector<uint8_t>& sourceBytes = *sourceBytesOpt;

    // The source entry's OWN current Font size/Volume -- used as the
    // "leave this field unchanged" value for groups 1/2 below.
    // kronos::FontSize's enum order matches the confirmed encoding value
    // directly (see its own doc comment in PcgFile.h), so this cast needs
    // no lookup table.
    const auto& sourceParams = pcg.setlists()[kSetlistIndex].songs[kSourceSlot].params;
    const int originalFontSizeValue = static_cast<int>(sourceParams.fontSize);
    const int originalVolume = sourceParams.volume;

    bool ok = true;

    // Group 1 (slots 10-14): Font size only, Volume left as the source's own.
    for (int i = 0; i < 5; ++i) {
        const std::string label = std::string("TEST FontSize: ") + kFontSizeNames[i];
        ok = writeSlot(pcg, 10 + i, sourceBytes, kFontSizeValues[i], originalVolume, -1, label, error) && ok;
    }

    // Group 2 (slots 15-19): Volume only, Font size left as the source's own.
    for (int i = 0; i < 5; ++i) {
        const std::string label = "TEST Volume: " + std::to_string(kVolumes[i]);
        ok = writeSlot(pcg, 15 + i, sourceBytes, originalFontSizeValue, kVolumes[i], -1, label, error) && ok;
    }

    // Group 3 (slots 20-24): both at once, the same 5 pairings from groups 1/2.
    for (int i = 0; i < 5; ++i) {
        const std::string label = std::string("TEST FontSize: ") + kFontSizeNames[i] + " Volume: " + std::to_string(kVolumes[i]);
        ok = writeSlot(pcg, 20 + i, sourceBytes, kFontSizeValues[i], kVolumes[i], -1, label, error) && ok;
    }

    // Group 4 (slots 25-29): Font size only, same wrap-probe Comment text in
    // every slot -- Volume left as the source's own, same as group 1.
    const std::string wrapTestText = makeWrapTestText();
    for (int i = 0; i < 5; ++i) {
        ok = writeSlot(pcg, 25 + i, sourceBytes, kFontSizeValues[i], originalVolume, -1, wrapTestText, error) && ok;
    }

    // Group 5 (slots 30-45): Color only, one slot per real Kronos color --
    // Font size/Volume left as the source's own.
    for (int i = 0; i < 16; ++i) {
        const std::string label = std::string("TEST Color: ") + kColorNames[i];
        ok = writeSlot(pcg, 30 + i, sourceBytes, originalFontSizeValue, originalVolume, i + 1, label, error) && ok;
    }

    if (!ok) {
        std::fprintf(stderr, "Failed writing a test slot: %s\n", error.c_str());
        return 1;
    }

    const std::string outputPath = testOutputPath(inputPath);
    if (!pcg.save(outputPath, error)) {
        std::fprintf(stderr, "Failed to save %s: %s\n", outputPath.c_str(), error.c_str());
        return 1;
    }

    std::printf("Wrote %s\n", outputPath.c_str());
    return 0;
}

File: tools/generate_setlist_test_matrix.js
// Hardware-validation helper -- NOT part of the app itself, never loaded by
// index.html. Generates a real .PCG/.SNG file with a matrix of Setlist Color/
// Volume/Comment/Font-size permutations, so the results can be checked by
// eye on a real Kronos, one slot at a time -- see STATE.md for the full
// rationale (this is how every confirmed byte offset in this project has
// been verified: real ground truth, not a plausible-looking guess).
//
// Deliberately NOT a from-scratch script re-implementing the SBK1 byte
// math -- it drives the real app's own bridge methods (getSongRecordBytes/
// putSongRecordBytes/saveFileAs) and the real JS codecs
// (setlist-editor-comment-and-font.js/setlist-editor-color.js/setlist-
// editor-volume.js). The whole point of this exercise is validating THAT
// code path against real hardware; a parallel implementation would only
// prove its own math right, not the app's.
//
// Usage:
//   1. Build and run the real native app (see README.md -- this needs the
//      actual CHOC/WebView bridge, mock_bridge.js's fakes won't produce a
//      loadable real file).
//   2. Open your minimal test file (one populated Setlist entry, everything
//      else empty -- see STATE.md's save-to-disk section for why a minimal
//      file is the right starting point) via the app's own Open dialog.
//   3. Open devtools (right-click -> Inspect Element; enabled via
//      options.enableDebugMode in main.cpp) and switch to the Console tab.
//   4. Paste this entire file's contents into the console and press enter
//      (defines the two functions below, does nothing else on its own).
//   5. Run:  await generateSetlistTestMatrix()
//      -- edits happen in memory only at this point, same as any other edit
//      in this app; nothing is written to disk yet.
//   6. Check the console output, then run:
//        await saveTestFile("/absolute/path/to/output.pcg")
//      -- writes the whole dataset (your one real entry PLUS the 15 new
//      test slots below, everything else untouched) to that path.
//   7. Load output.pcg onto your Kronos and check slots 010-045 by eye.
//
// Both functions default to SET LIST INDEX 0, SOURCE SLOT 0 (Kronos's own
// 000-127 numbering) and the MOST RECENTLY OPENED dataset -- edit the
// constants below first if your test file's one real entry lives somewhere
// else.

const TEST_MATRIX_SETLIST_INDEX = 0;
const TEST_MATRIX_SOURCE_SLOT = 0;

// The 5 confirmed Font size values (docs/content/format/index.md §4.4) and 5 representative
// Volume values (0/1 = the extremes past "silent", 10/100 = two ordinary
// values, 127 = max) -- one each per group, in this fixed order.
const TEST_MATRIX_FONT_SIZES = ["XS", "S", "M", "L", "XL"];
const TEST_MATRIX_VOLUMES = [0, 1, 10, 100, 127];

// Group 4's word-wrap probe: sequential 2-digit tokens ("01 02 03 ... 80"),
// same text in all 5 Font sizes so wherever the real hardware breaks each
// line can be read straight off the screen and reported back exactly (e.g.
// "line 1 ends at 09, line 2 at 19") -- no ambiguity about which word is
// which, unlike a real sentence. See frontend/readme-screen.txt's own
// (unverified, likely fabricated) claims about Kronos text rendering --
// this is how to actually find out, not guess.
const TEST_MATRIX_WRAP_TEXT = Array.from({ length: 80 }, (_, i) => String(i + 1).padStart(2, "0")).join(" ");

// The 16 real Kronos Set List colors, confirmed against real hardware via
// this very group -- see pane.js's SETLIST_COLOR_NAMES (kept in sync) and
// STATE.md for the confirmation record.
const TEST_MATRIX_COLOR_NAMES = [
  "Default", "Charcoal", "Brick", "Burgundy", "Ivy", "Olive", "Gold", "Cacao",
  "Indigo", "Navy", "Rose", "Lavender", "Azure", "Denim", "Silver", "Slate",
];

async function generateSetlistTestMatrix() {
  const datasets = await window.listDatasets();
  if (datasets.length === 0) throw new Error("No dataset open -- open your test file first.");
  const datasetId = datasets[datasets.length - 1].datasetId;
  console.log(`Using dataset ${datasetId} (${datasets[datasets.length - 1].displayName}), Set List ${TEST_MATRIX_SETLIST_INDEX}.`);

  // Relative to the loaded PAGE (frontend/index.html), not to this script's
  // own path on disk -- code pasted into devtools console runs in the
  // page's own context, so a relative import() resolves the same way
  // pane-setlist-editor.js's own dynamic import of these same files already does.
  const { decodeSetlistComment, encodeSetlistComment } = await import("./components/kronos/setlist-editor-comment-and-font.js");
  const { encodeSlotColor } = await import("./components/kronos/setlist-editor-color.js");
  const { encodeSlotVolume } = await import("./components/kronos/setlist-editor-volume.js");

  const sourceResult = await window.getSongRecordBytes(datasetId, TEST_MATRIX_SETLIST_INDEX, TEST_MATRIX_SOURCE_SLOT);
  if (!sourceResult.ok) throw new Error(`Couldn't read the source slot: ${sourceResult.error}`);
  const sourceBytes = Uint8Array.from(sourceResult.bytes);
  const sourceDecoded = decodeSetlistComment(sourceBytes);

  // Writes `sourceBytes` into `slot`, with `fontSize`/`volume`/`color`
  // overridden where given (null/undefined leaves the source entry's own
  // value untouched) and Comment set to `label` -- mirrors exactly what
  // pane.js's buildColorSection()/buildCommentSection()/buildVolumeSection()
  // already do per edit, just scripted across many slots instead of one
  // click at a time.
  async function writeSlot(slot, { fontSize, volume, color, label }) {
    let bytes = encodeSetlistComment(sourceBytes, {
      comment: label,
      fontSize: fontSize ?? sourceDecoded.fontSize,
    });
    if (volume != null) bytes = encodeSlotVolume(bytes, volume);
    if (color != null) bytes = encodeSlotColor(bytes, color);

    const result = await window.putSongRecordBytes(datasetId, TEST_MATRIX_SETLIST_INDEX, slot, Array.from(bytes));
    if (!result.ok) throw new Error(`Slot ${slot}: ${result.error}`);
    console.log(`Slot ${slot}: ${label}`);
  }

  // Group 1 (slots 10-14): Font size only, Volume left as the source's own.
  for (let i = 0; i < TEST_MATRIX_FONT_SIZES.length; i++) {
    const fontSize = TEST_MATRIX_FONT_SIZES[i];
    await writeSlot(10 + i, { fontSize, volume: null, label: `TEST FontSize: ${fontSize}` });
  }

  // Group 2 (slots 15-19): Volume only, Font size left as the source's own.
  for (let i = 0; i < TEST_MATRIX_VOLUMES.length; i++) {
    const volume = TEST_MATRIX_VOLUMES[i];
    await writeSlot(15 + i, { fontSize: null, volume, label: `TEST Volume: ${volume}` });
  }

  // Group 3 (slots 20-24): both at once, the same 5 pairings from groups 1/2.
  for (let i = 0; i < TEST_MATRIX_FONT_SIZES.length; i++) {
    const fontSize = TEST_MATRIX_FONT_SIZES[i];
    const volume = TEST_MATRIX_VOLUMES[i];
    await writeSlot(20 + i, { fontSize, volume, label: `TEST FontSize: ${fontSize} Volume: ${volume}` });
  }

  // Group 4 (slots 25-29): Font size only, same wrap-probe Comment text in
  // every slot -- Volume left as the source's own, same as group 1.
  for (let i = 0; i < TEST_MATRIX_FONT_SIZES.length; i++) {
    const fontSize = TEST_MATRIX_FONT_SIZES[i];
    await writeSlot(25 + i, { fontSize, volume: null, label: TEST_MATRIX_WRAP_TEXT });
  }

  // Group 5 (slots 30-45): Color only, one slot per real Kronos color, in
  // TEST_MATRIX_COLOR_NAMES' order -- Font size/Volume left as the source's
  // own.
  for (let i = 0; i < TEST_MATRIX_COLOR_NAMES.length; i++) {
    const color = i + 1;  // SlotParams.color is 1-based
    const name = TEST_MATRIX_COLOR_NAMES[i];
    await writeSlot(30 + i, { fontSize: null, volume: null, color, label: `TEST Color: ${name}` });
  }

  console.log("Done -- 36 slots written (010-045). Now run: await saveTestFile(\"/absolute/path/to/output.pcg\")");
}

async function saveTestFile(path) {
  const datasets = await window.listDatasets();
  if (datasets.length === 0) throw new Error("No dataset open.");
  const datasetId = datasets[datasets.length - 1].datasetId;
  const result = await window.saveFileAs(datasetId, path);
  if (!result.ok) throw new Error(result.error);
  console.log(`Saved to ${path}`);
}
